dep-types-inference/dep-types.md

```diff
@@ -0,0 +1,687 @@
+
+We can construct the program. Before we're doing inference, we need an auxillary mechanism. It whill compare two types and either find a set of substitutions to convert both to the same type, or bail out with error if that isn't possible. This mechanism is called unification.
+
+Usially, in Hindley-Milner inference, the type of unifier is
+
+```haskell
+unify :: Type -> Type -> Either UnificationError Subst
+```
+
+... and the substitution produced on success is applied to all related types that caller uses later. This makes it prone to error, because you could forget to apply some substitution `s7` and this will lead to rather hard time debugging the inference engine.
+
+I'd suggest hiding the accumulated substitutions as `State` inside some monad, and split the unifier onto the part that does the job and the part that applies bindings and adds the resulting ones to the state. Since all the names are unique, the substitutions can never become obsolete.
+
+Please keep in mind that since we have _dependent_ types, the types can and will contain expressions interleaved with types. Also, any value that got to the unifier will be normalized, which means some cases should not be possible.
+
+Lets take apart working half of the unifier:
+```haskell
+unify
+  :: (Refreshes m, HasSubst m, CanThrow UnificationError m)
+  => AST
+  -> AST
+  -> m ()
+```
+We also need to throw errors, sometimes.
+
+The first part is unification of two free variables:
+```haskell
+unify (Var n) (Var m)
+  | n == m    = mempty
+  | otherwise = modify (FreeVar n ==> (Var m) <>)
+```
+Which is just replacing one with the another.
+
+The next case to check is if only one side is a free variable:
+```haskell
+unify (Var n) m
+  | occurs n m = die $ Occurs n m
+  | otherwise  = modify (FreeVar n ==> m <>)
+
+unify m (Var n)
+  | occurs n m = die $ Occurs n m
+  | otherwise  = modify (FreeVar n ==> m <>)
+```
+We use the occurence checker from [#Substitution](#Substitution) to check if the resulting type does not refer to itself. How to combine dependent and recursive types is pretty open question.
+
+A bound variable can only unify with itself. It represents some value or type that is not known yet - or can vary - in this region of the program.
+```haskell
+unify (Bound t) (Bound u) | t == u = mempty
+```
+
+For some structures unification is just a descent.
+```haskell
+unify (App a b) (App c d) = do
+  unify a c
+  unify b d
+
+unify Star Star = do
+  return ()
+
+unify (Lit l) (Lit k)
+  | l == k = mempty
+```
+
+For functions and their types though, it is a bit more complicated process. We need to unify the type of the argument and then replace the argument in the body of function (or dependent function type) with a name from its counterpart and only then we can unify it.
+```haskell
+unify (Pi n t b) (Pi m u c) = do
+  unify t u
+  unify b (subst (Bound m ==> Rigid n) c)
+
+unify (Lam n t b) (Lam m u c) = do
+  unify t u
+  unify b (subst (Bound m ==> Rigid n) c)
+```
```

Comment:
why do you repeat `unify (Lam n t b) (Lam m u c) = do` twice?

Reply:
```haskell
unify (Pi n t b) (Pi m u c) = do
  unify t u
  unify b (subst (Bound m ==> Rigid n) c)
unify (Lam n t b) (Lam m u c) = do
  unify t u
  unify b (subst (Bound m ==> Rigid n) c)
```

Where? The first clause is for `Pi`-types.

Reply:
They just appear to have the same unification procedure.